### PR TITLE
OEP(volume resize): add proposal for cStor volume resize

### PR DESCRIPTION
**Pull request by mittachaitu**
This commit adds an OpenEBS Enhancement Proposal to define the
specifications and design required to implement resize on cStor volumes.

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

### CODE REVIEW

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    * [High Level Design](#high-level-design)
+      * [CStorVolumeClaim -- existing custom resource](#cstorvolumeclaim----existing-custom-resource)
+      * [CStorVolume -- existing custom resource](#cstorvolume----existing-custom-resource)
+      * [CStorVolumeReplica -- existing custom resource](#cstorvolumereplica----existing-custom-resource)
+* [Drawbacks](#drawbacks)
+* [Alternatives](#alternatives)
+
+## Summary
+
+This proposal charts out the design details to implement resize workflow on cStor Volumes.
```
Comment (kmova) [inline contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: Is this applicable for only CSI driver based cStor volumes? or does it work for both non-csi and the new ones?

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+## Motivation
+
+Resizing volume is the most prominent feature for any storage providers.
+In this we will concentrate only on cStor volumes.
+
+### Goals
+
+- Ability to resize a cStor volume in kubernetes cluster by taking trigger
```
Comment (kmova) [inline contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: Reword this as `Ability to resize a cStor volume by changing the capacity in the PVC spec.`

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+      * [CStorVolume -- existing custom resource](#cstorvolume----existing-custom-resource)
+      * [CStorVolumeReplica -- existing custom resource](#cstorvolumereplica----existing-custom-resource)
+* [Drawbacks](#drawbacks)
+* [Alternatives](#alternatives)
+
+## Summary
+
+This proposal charts out the design details to implement resize workflow on cStor Volumes.
+
+## Motivation
```
Comment (kmova) [inline contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: You can remove this section.

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+- Ability to resize a cStor volume in kubernetes cluster by taking trigger
+  point as pvc spec capacity change.
+
+## Proposal
+
+### User Stories
+
+#### Resize cStor Volume:
+As an application developer I should be able to resize a volume on
```
Comment (kmova) [inline contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: on the fly means that it can be resized while the application pods is still running. If this doesn't support online resizing, we have explicitly mention it.

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+fly(when application consuming volume).
+
+### Implementation Details
+
+There was a working flow for resizing cStor volumes via REST and grpc calls.
```
Comment (kmova) [inline contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: Are you referring to the steps that can be used via REST/gRPC for non-csi cstor volumes? What are the API and how can users invoke them?

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+Since there are more visible barriers(mainly when container or pod restarts)
+REST or grpc calls will fail.
+
+The sections below can be assumed to be specific to `cStor` unless mentioned
+otherwise.
+
+### Custom Resources used to resize cStor volume
```
Comment (kmova) [inline contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: Prior to mentioning these YAMLs, provide a step on what triggers the resize workflow.

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+  cstor-volume-mgmt.
+  - cstor-istgt will disconnect the replicas from which we have not received the
+    response and then the successful response will send to the sidecar.
+    cstor-istgt will send updated size to replicas at the time of management
+    connection with the target.
+  Note: Processing `istgtcontrol resize` is a blocking call.
+
+- zfs will receives the resize request, it will resize the corresponding volume
+  and sent back the response to cstor-istgt(aka target).
+  Note:
```
Comment (kmova) [inline contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: Push these under the limitations section, if currently not supported.

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    In ephemeral case(temporary storage) when resize is success at the time of
+    node unavailable cases and when the node up need to take care of rebuilding
+    snapshots of different sizes.
+
+## High Level cStor Volume Resize Workflow
```
Comment (kmova) [inline contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: See if the diagram in the link can be converted to ascii drawing and inserted into this section.

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md (hunk 4 above)
Comment (mittachaitu) [reply contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: We are supporting online as well as offline.

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md (hunk 5 above)
Comment (mittachaitu) [reply contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: There is no support for Non-CSI cStor volume resize(PR was not merged because of pitfalls). I will remove this.

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md (hunk 7 above)
Comment (mittachaitu) [reply contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: kiran this document is outdated. Updated with the latest information.

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md (hunk 1 above)
Comment (mittachaitu) [reply contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: Updated informing this is only for CSI cStor volumes.

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md (hunk 2 above)
Comment (mittachaitu) [reply contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: Done

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md (hunk 3 above)
Comment (mittachaitu) [reply contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: Removed

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md (hunk 6 above)
Comment (mittachaitu) [reply contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: Added step what users need to do.

File: contribute/design/1.x/maya/20190107-cStor-volume-resize.md (hunk 8 above)
Comment (mittachaitu) [reply contribute/design/1.x/maya/20190107-cStor-volume-resize.md]: Updated with the relative image link(It will load on the page itself). PTAL on https://github.com/mittachaitu/openebs/blob/resize_oep/contribute/design/1.x/csi/20190107-cStor-volume-resize.md#high-level-cstor-volume-resize-workflow

File: contribute/design/1.x/csi/20190107-cStor-volume-resize.md [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+```
+In the above PVC spec capacity is updated to 10Gi but status capacity is 5Gi(Which
+means resize is in progress).
+
+### Implementation Details
+
+The sections below can be assumed to be specific to `cStor` unless mentioned
+otherwise.
+
+### Custom Resources used to resize cStor volume
```
Comment (kmova) [inline contribute/design/1.x/csi/20190107-cStor-volume-resize.md]: This block appears out of no-where. Move this section below the implementation workflow. And create a link to this section, in the note that is mentioning to refer to this.

File: contribute/design/1.x/csi/20190107-cStor-volume-resize.md [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+    autoUpgrade: false
+    desired: 1.4.0
+    status:
+      current: 1.4.0
+      dependentsUpgraded: true
+      lastUpdateTime: null
+      state: ""
+```
+
+Sample cstorvolume yaml when resize is in progress
```
Comment (kmova) [inline contribute/design/1.x/csi/20190107-cStor-volume-resize.md]: Similar to above comment. Move this section below the implementation workflow. And create a link to this section, in the note that is mentioning to refer to this.

File: contribute/design/1.x/csi/20190107-cStor-volume-resize.md [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+  Note: Processing `istgtcontrol resize` is a blocking call.
+
+- zfs will receives the resize request, it will resize the corresponding volume
+  and sent back the response to cstor-istgt(aka target).
+
+Once resize operation is succeed at OpenEBS side CSI node plugin(/kubelet) will
+trigger resize operation on filesystem level as part of online resizing.
+
+## High Level cStor Volume Resize Workflow
+  ![Resize Workflow](Resize_work_flow.png)
```
Comment (kmova) [inline contribute/design/1.x/csi/20190107-cStor-volume-resize.md]: add relative path like - [./Resize_work_flow.png]

File: contribute/design/1.x/csi/20190107-cStor-volume-resize.md (hunk 9 above)
Comment (mittachaitu) [reply contribute/design/1.x/csi/20190107-cStor-volume-resize.md]: Done

File: contribute/design/1.x/csi/20190107-cStor-volume-resize.md (hunk 10 above)
Comment (mittachaitu) [reply contribute/design/1.x/csi/20190107-cStor-volume-resize.md]: Done

File: contribute/design/1.x/csi/20190107-cStor-volume-resize.md (hunk 11 above)
Comment (mittachaitu) [reply contribute/design/1.x/csi/20190107-cStor-volume-resize.md]: Done!! Thanks for review comments @kmova .
